scripts/py_matter_idl/matter_idl/data_model_xml_parser.py

```diff
@@ -0,0 +1,172 @@
+#!/usr/bin/env python3
+
+# Copyright (c) 2022 Project CHIP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import os
+import sys
+
+import click
+
+try:
+    from matter_idl.data_model_xml import ParseSource, ParseXmls
+except ImportError:
+    sys.path.append(os.path.abspath(
+        os.path.join(os.path.dirname(__file__), '..')))
+    from matter_idl.data_model_xml import ParseSource, ParseXmls
+
+from matter_idl.generators import GeneratorStorage
+from matter_idl.generators.idl import IdlGenerator
+from matter_idl.matter_idl_parser import CreateParser
+from matter_idl.matter_idl_types import Idl
+
+
+class InMemoryStorage(GeneratorStorage):
+    def __init__(self):
+        super().__init__()
+        self.content: Optional[str] = None
+
+    def get_existing_data(self, relative_path: str):
+        # Force re-generation each time
+        return None
+
+    def write_new_data(self, relative_path: str, content: str):
+        if self.content:
+            raise Exception(
+                "Unexpected extra data: single file generation expected")
+        self.content = content
+
+
+def normalize_order(idl: Idl):
+    """Re-sorts contents of things inside a cluster so that
```

Comment:
Does this happen much where things are not in the same order? I don't doubt it, but it probably points to some spec weirdness.

Reply:
I don't think it is easily solvable: the order comes from .matter IDL generated by zap, and zap generally does a SQL select ... to remove non-determinism I believe those selects have `ORDER BY` but then it becomes tedious for python code to guess what sort order (usually ID, but sometimes name ... or for request/response even different) it should use.

I found it much easier to just say "stable sorting for everyone"

Reply:
Added a comment to describe this.